lib/SILOptimizer/Utils/SILInliner.cpp

```diff
@@ -18,13 +18,181 @@
 using namespace swift;
 
 bool SILInliner::canInlineFunction(FullApplySite AI) {
-  // For now, we cannot inline begin_apply at all.
-  if (isa<BeginApplyInst>(AI))
-    return false;
-
   return AI.getFunction() != &Original;
 }
 
+/// Utility class for rewiring control-flow of inlined begin_apply functions.
+class BeginApplySite {
+  SmallVector<SILBasicBlock *, 4> ExitingBlocks;
+  SmallVector<AllocStackInst*, 8> YieldedIndirectValues;
+  SILLocation Loc;
+  SILBuilder &Builder;
+  BeginApplyInst *BeginApply;
+  SILFunction *F;
+  EndApplyInst *EndApply = nullptr;
+  SILBasicBlock *EndApplyBB = nullptr;
+  SILBasicBlock *EndApplyBBMerge = nullptr;
+  AbortApplyInst *AbortApply = nullptr;
+  SILBasicBlock *AbortApplyBB = nullptr;
+  SILBasicBlock *AbortApplyBBMerge = nullptr;
+  SILArgument *IntToken = nullptr;
+
+  unsigned YieldNum = 0;
+  SmallVector<SILBasicBlock*, 8> YieldResumes;
+  SmallVector<SILBasicBlock*, 8> YieldUnwinds;
+
+  void
+  getYieldCaseBBs(SmallVectorImpl<std::pair<SILValue, SILBasicBlock *>> &Result,
+                  SmallVectorImpl<SILBasicBlock *> &Dests) {
+    unsigned Token = 0;
+    for (auto *Blk : Dests) {
+      Result.push_back(std::make_pair(
+          SILValue(Builder.createIntegerLiteral(
+              Loc,
+              SILType::getBuiltinIntegerType(
+                  32, Builder.getFunction().getModule().getASTContext()),
+              Token++)),
+          Blk));
+    }
+  }
+
+public:
+  BeginApplySite(BeginApplyInst *BeginApply, SILLocation Loc,
+                 SILBuilder &Builder)
+      : Loc(Loc), Builder(Builder), BeginApply(BeginApply),
+        F(BeginApply->getFunction()) {}
+
+  static Optional<BeginApplySite> isa(FullApplySite AI, SILLocation Loc,
+                                      SILBuilder &Builder) {
+    auto *BeginApply = dyn_cast<BeginApplyInst>(AI);
+    if (!BeginApply)
+      return None;
+    return BeginApplySite(BeginApply, Loc, Builder);
+  }
+
+  void collectCallerExitingBlocks() {
+    F->findExitingBlocks(ExitingBlocks);
+  }
+
+  void processApply(SILBasicBlock *ReturnToBB) {
+    // Handle direct and indirect results.
+    for (auto YieldedValue : BeginApply->getYieldedValues()) {
+      // Insert an alloc_stack for indirect results.
+      if (YieldedValue->getType().isAddress()) {
```

Comment:
Is this just a placeholder?  Because this shouldn't be necessary — the indirect yield comes from a value that should be live after inlining —  and it's not semantically valid to introduce a copy if e.g. the yielded value is `inout`.

Reply:
I wrote this under the assumption we yield indirect values only ``@in``.

I agree. I could have just used the inlined alloc_stack instead.

I am not sure how inout yields work. What is being yielded on the callee side?  I would assume the following: We create a temporary location in the callee that is then replaced by the caller's location if we inline. Is the following a correct use of an inout yield?

Caller:
```
%caller_loc = alloc_stack
copy_addr %something to [initialization] %caller_loc
(%caller_loc, %token)= begin_apply %coroutine()
```

Callee:

```
sil [transparent] @coroutine : $@yield_once() -> (@yields @inout Something) {
 entry:
   %temp = alloc_stack $Something
   copy_addr %something to %temp
   yield %temp : $*Something, resume resume, unwind unwind

resume:
   apply %some_use(%temp)
```

Couldn't this just as well be expressed as a parameter that we pass inout to the begin_apply?

Reply:
Oh, right the interesting case is this:

Caller:

```
(%caller_loc, %token)= begin_apply %coroutine()
copy_addr %something to %caller_loc // This is now available on the resume path of the callee.
```

Reply:
You should think of the `yield` like a function call.  The coroutine can produce any address it wants as the `yield` value, and the caller just receives an opaque pointer — just like a function can pass any address it wants as an `inout` argument, and the callee just receives an opaque pointer.  The convention on the yielded address is what the caller is expected to have done before it resumes the coroutine, so e.g. if the caller wants to save a value that was yielded `@in` to it, it's required to move the value into caller-owned memory.  (In the inliner, you can just assume that's handled correctly; code-motion / copy-propagation optimizations will have to worry about it.)

Reply:
Okay, thanks.

Fixed both issues in https://github.com/apple/swift/pull/18477